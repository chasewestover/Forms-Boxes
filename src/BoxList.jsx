import { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import {v4 as uuid} from 'uuid';

function BoxList() {
  const [boxList, setBoxList] = useState([]);

  function removeBox(box){
    setBoxList(blist => blist.filter(b => b !== box));
  }

  function addBox(box) {
    box.key = uuid()
    setBoxList(blist => [...blist, box]);
  }

  return (
    <div>
      <NewBoxForm addBox={addBox} />
      <ul>
      {boxList.map(box => (
        <Box
          height={box.height}
          width={box.width}
          backgroundColor={box.backgroundColor}
          removeBox={() => removeBox(box)}/>
      ))}
      </ul>
    </div>
  );
}

export default BoxList;