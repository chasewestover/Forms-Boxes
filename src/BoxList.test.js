import { render, fireEvent  } from '@testing-library/react';

import BoxList from "./BoxList";


describe(" test BoxList renders correctly", function(){

  test("render BoxList", function(){
    render(<BoxList />)
  })
  test("BoxList snapshot test", function(){
    const {container} = render(<BoxList/>)
    expect(container).toMatchSnapshot();
  });
})

describe("entering box data into form", function(){
  test("adding a new box", function(){
    const {getByLabelText, queryByText, container, debug} = render(<BoxList/>);
    debug();
    expect(container.querySelector(".Box-box")).not.toBeInTheDocument();
    expect(container.querySelector(".Box-box")).toEqual(null);

    const heightInput = getByLabelText("height:");
    const widthInput = getByLabelText("width:");
    const backgroundColorInput = getByLabelText("background color:");
    const submitBtn = queryByText("Submit");

    fireEvent.change(heightInput, {target: {value: "200"}});
    fireEvent.change(widthInput, {target: {value: "200"}});
    fireEvent.change(backgroundColorInput, {target: {value: "green"}});
    fireEvent.click(submitBtn);

    const box = container.querySelector(".Box-box");
    expect(box).toBeInTheDocument();
  });
});

describe("tests validation of form", function(){
  test("check form data is numeric, won't render box with bad data",function(){

    const {getByLabelText, queryByText, container, debug} = render(<BoxList/>);

    const heightInput = getByLabelText("height:");
    const widthInput = getByLabelText("width:");
    const backgroundColorInput = getByLabelText("background color:");
    const submitBtn = queryByText("Submit");


    fireEvent.change(heightInput, {target: {value: "two hundred"}});
    fireEvent.change(widthInput, {target: {value: "one hundred"}});
    fireEvent.change(backgroundColorInput, {target: {value: "green"}});
    fireEvent.click(submitBtn);

    const box = container.querySelector(".Box-box");
    expect(box).not.toBeInTheDocument();
    expect(container.querySelector(".Box-error")).toBeInTheDocument();

  
  });
})