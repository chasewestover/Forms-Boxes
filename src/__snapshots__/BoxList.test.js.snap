// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` test BoxList renders correctly BoxList snapshot test 1`] = `
<div>
  <div>
    <div>
      <form>
        <label
          for="height"
        >
          height:
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
        <label
          for="width"
        >
          width:
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
        <label
          for="backgroundColor"
        >
          background color:
        </label>
        <input
          id="backgroundColor"
          name="backgroundColor"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
